### ISSUE TITLE
Pickle issue with private members in storemodules

### ISSUE THREAD

**Issue by Booplicate** [renpy/renpy#4024]
Using the following code to define a custom statement
```renpy
python early in test_store:
    from collections import namedtuple

    __TestNT = namedtuple("__TestNT", ("param1", "param2"))

    def __parse_test_stmt(lex):
        lex.expect_noblock("test_stmt")
        lex.expect_eol()
        lex.advance()

        return __TestNT("1", "2")

    def __execute_test_stmt(parsed_data):
        pass

    def __predict_test_stmt(parsed_data):
        pass

    def __lint_test_stmt(parsed_data):
        pass

    renpy.register_statement(
        "test_stmt",
        parse=__parse_test_stmt,
        execute=__execute_test_stmt,
        predict=__predict_test_stmt,
        lint=__lint_test_stmt
    )
```
And then this code in a test `rpy`
```renpy
label test_label():
    test_stmt
    return
```
I'm getting this error
```
PicklingError: Can't pickle <class 'store.test_store.__TestNT'>: attribute lookup __TestNT on store.test_store failed
```
Happens when using renpy 8.1, on much older 7.x everything is fine, so might be a py3 thing.

**Comment by mal** [comment]
Try it without using double underscores for the prefix. That triggers [name mangling](https://docs.python.org/3/tutorial/classes.html#private-variables) in Python which is likely the problem.

**Comment by Booplicate** [comment]
I know that the issue *appears* because I made the class private. I'm also aware of Python name mangling and that RenPy has its own version of name mangling (which might be the *actual root*). Nevertheless, I think this should be fixed, since this code worked before and RenPy has no documentation regarding avoiding using private variables.

**Comment by Booplicate** [comment]
Generally private variables work in RenPy, but in this case of a custom statement, they don't.

**Comment by Gouvernathor** [comment]
`__TestNT = namedtuple("__TestNT", ("param1", "param2"))`
The variable name is mangled in renpy (in python it wouldn't, since it's not a class/instance attribute but a global variable), but the string given to namedtuple is not, because it's a string. This makes the class not being saved under its own name and original module, which breaks pickling.
I don't know how that could have worked under py2.

**Comment by Booplicate** [comment]
> This makes the class not being saved under its own name and original module, which breaks pickling

Okay, that makes more sense! I even tried to see if there's any difference and launched both py2 and py3 (I had to comment out the line where I used it, but it was still defined) versions, they looked the same.

Then I got another question, while what you say seems to be correct, why does it happen in the first place here? Private variables always worked and are working in RenPy. Why is in this one case RenPy/pickle trying to access a private member from the outer scope? Or if they always do it, why is this case not working? For example, it has no problems with me using private functions there.

Perhaps something has been changed somewhere. Was it pickle, RenPy script parser, or maybe it's the AST?

**Comment by Gouvernathor** [comment]
This has to do with how pickle works with functions and classes : it basically registers the name (accessed by `__name__`) and module name, and tries to resolve `(modname).(name)` when unpickling, raising an error if it doesn't exist.
The particularity of namedtuple is that unlike normal class or function definition, you have to input the class's `__name__` as a string. For `class __A:`, `__name__` will be `_garbage__A` and the variable name (containiln the type object) will have the same name. Same for `def __f():` with `_garbage__f`. But with that namedtuple, `__name__` is the string you've input, "__TestNT", and the varname containing the type object is `_garbage__TestNT`.

Last thing, most functions usually avoid being picled by not having anything refer to them, so if they're pickle-unreachable it's fine. But any object instance will by nature keep a reference to its class, so every type whose instances are intended to be pickled need to be available under their original name.

**Comment by Booplicate** [comment]
Hmm, yeah that all makes sense, maybe RenPy pickles (caches) AST nodes, in which case it might keep references to the parsed data for which I used my namedtuple, then RenPy/pickle got incorrect name because of mangling.

I tried to mangle the name using the default renpy function and my nametuple worked like a charm. So I guess the issue was kind of resolved. Interesting how it works on py2, though.
